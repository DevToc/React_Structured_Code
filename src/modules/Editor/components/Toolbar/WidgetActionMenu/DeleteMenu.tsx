import { IconButton, Tooltip } from '@chakra-ui/react';

import { setActiveWidget } from 'modules/Editor/store/store';
import { selectActiveWidgetIds } from 'modules/Editor/store/widgetSelector';
import { removeWidget } from 'modules/Editor/store/infographSlice';
import { selectActivePage } from 'modules/Editor/store/pageSelector';
import { useAppDispatch, useAppSelector } from 'modules/Editor/store/hooks';

import { ReactComponent as TrashIcon } from 'assets/icons/trash.svg';

export const DeleteMenu = () => {
  const dispatch = useAppDispatch();

  const activePageId = useAppSelector(selectActivePage);
  const activeWidgetIds = useAppSelector(selectActiveWidgetIds);

  const isDisabled = !activeWidgetIds.length;
  const label = 'Delete widget';

  const removeActiveWidgets = () => {
    dispatch(removeWidget({ pageId: activePageId, widgetIds: activeWidgetIds }));
    dispatch(setActiveWidget([]));
  };

  return (
    <Tooltip hasArrow placement='bottom' label={label} bg='black'>
      <IconButton
        isDisabled={isDisabled}
        size='sm'
        aria-label={label}
        icon={<TrashIcon />}
        onClick={removeActiveWidgets}
      />
    </Tooltip>
  );
};
