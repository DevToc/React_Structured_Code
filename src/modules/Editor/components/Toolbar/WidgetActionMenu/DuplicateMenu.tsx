import { IconButton, Tooltip } from '@chakra-ui/react';

import { TEXT_CREATED } from 'constants/mixpanel';
import { Mixpanel } from 'libs/third-party/Mixpanel/mixpanel';
import { TextWidgetData } from 'widgets/TextBasedWidgets/TextWidget/TextWidget.types';
import { getWidgetTypeFromId } from 'widgets/Widget.helpers';
import { WidgetType } from 'types/widget.types';
import { ReactComponent as CopyIcon } from 'assets/icons/copy.svg';

import { selectActiveWidgets } from 'modules/Editor/store/widgetSelector';
import { selectActivePage } from 'modules/Editor/store/pageSelector';
import { selectWidgets } from 'modules/Editor/store/infographSelector';
import { addNewWidget, useAppDispatch, useAppSelector } from 'modules/Editor/store';
import { Keyboard } from 'modules/common/components/Keyboard';

import { createCopyWidgets, parseCopyWidgets } from '../../Clipboard';

export const DuplicateMenu = () => {
  const dispatch = useAppDispatch();

  const activePageId = useAppSelector(selectActivePage);
  const activeWidgets = useAppSelector(selectActiveWidgets);
  const widgets = useAppSelector(selectWidgets);

  const isDisabled = !activeWidgets.length;
  const label = 'Duplicate';

  const onDuplicate = () => {
    const copyWidgets = createCopyWidgets(activeWidgets, widgets, activePageId);
    const newWidgets = parseCopyWidgets(copyWidgets, activePageId);

    dispatch(addNewWidget(newWidgets));

    if (!activeWidgets[1] && activeWidgets[0]?.id) {
      const widgetType = getWidgetTypeFromId(activeWidgets[0]?.id);
      if (widgetType === WidgetType.Text) {
        const { textTag } = widgets[activeWidgets[0]?.id] as TextWidgetData;
        const textType = textTag || 'Other';

        Mixpanel.track(TEXT_CREATED, {
          from: 'Duplication',
          text_style: textType.replace(textType[0], textType[0].toUpperCase()),
        });
      }
    }
  };

  return (
    <Tooltip
      hasArrow
      placement='bottom'
      label={
        <>
          {label} <Keyboard shortcut={['ctrl', 'D']} macOs={['âŒ˜', 'D']} />
        </>
      }
      bg='black'
    >
      <IconButton isDisabled={isDisabled} size='sm' aria-label={label} icon={<CopyIcon />} onClick={onDuplicate} />
    </Tooltip>
  );
};
