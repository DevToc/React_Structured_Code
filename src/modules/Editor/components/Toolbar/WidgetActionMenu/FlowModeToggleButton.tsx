import { IconButton, Tooltip } from '@chakra-ui/react';
import { useCallback } from 'react';

import { ReactComponent as CubeTransparent } from 'assets/icons/cube_transparent.svg';
import { useFlowCoreValue, useSetFlowCoreState } from 'modules/Editor/components/FlowCore';

// NOTE: [JB] This component is for rendering the temporary button to toggle flow mode. It will be removed
export const FlowModeToggleButton = () => {
  const isEnabled = useFlowCoreValue('isEnabled');
  const setIsEnabled = useSetFlowCoreState('setIsEnabled');

  const label = `[BETA] Toggle Flow Mode - ${isEnabled ? 'ON' : 'OFF'}`;

  const toggleFlowMode = useCallback(() => {
    setIsEnabled(!isEnabled);
  }, [setIsEnabled, isEnabled]);

  return (
    <Tooltip hasArrow placement='bottom' label={label} bg='black'>
      <IconButton
        data-testid={'flowmode-toggle-button'}
        size='sm'
        aria-label={label}
        backgroundColor={isEnabled ? 'blue.100' : 'gray.50'}
        _hover={{
          backgroundColor: 'blue.100',
        }}
        _active={{
          backgroundColor: 'blue.200',
        }}
        color={isEnabled ? 'red.600' : 'gray.400'}
        icon={<CubeTransparent />}
        onClick={toggleFlowMode}
      />
    </Tooltip>
  );
};
