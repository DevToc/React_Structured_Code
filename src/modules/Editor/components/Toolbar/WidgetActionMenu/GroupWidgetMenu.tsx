import { Button, Tooltip } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';

import { useAppDispatch } from 'modules/Editor/store/hooks';
import { groupSelectedWidgets, unGroupSelectedWidgets } from 'modules/Editor/store';
import { selectActiveWidgets } from 'modules/Editor/store/widgetSelector';
import { useAppSelector } from 'modules/Editor/store/hooks';
import { Keyboard } from 'modules/common/components/Keyboard';

const GroupWidgetMenu = () => {
  const activeWidgets = useAppSelector(selectActiveWidgets);
  const dispatch = useAppDispatch();
  const { t } = useTranslation('editor_toolbar');

  const groupText = t('group.group');
  const unGroupText = t('group.ungroup');

  const singleGroupSelected =
    activeWidgets.length === 1 && activeWidgets[0].groupMembers.length > 0 && !!activeWidgets[0].groupId;
  const hasUngroup = singleGroupSelected;
  const hasGroup = activeWidgets.length > 1;

  const onGroup = () => dispatch(groupSelectedWidgets(activeWidgets));
  const onUnGroup = () => {
    const groupId = activeWidgets[0]?.groupId;
    if (groupId) dispatch(unGroupSelectedWidgets(groupId));
  };

  const onClick = () => {
    if (hasUngroup) return onUnGroup();
    if (hasGroup) return onGroup();
  };

  const isDisabled = !hasGroup && !hasUngroup;
  const label = hasGroup ? groupText : hasUngroup ? unGroupText : groupText;

  return (
    <Tooltip
      hasArrow
      placement='bottom'
      label={
        <>
          {label} <Keyboard shortcut={['ctrl', 'alt', 'G']} macOs={['⌘', '⌥', 'G']} />
        </>
      }
      bg='black'
    >
      <Button
        variant='ghost'
        fontWeight='semibold'
        data-testid='group-widget-menu'
        isDisabled={isDisabled}
        onClick={onClick}
        size='sm'
      >
        {label}
      </Button>
    </Tooltip>
  );
};

export { GroupWidgetMenu };
