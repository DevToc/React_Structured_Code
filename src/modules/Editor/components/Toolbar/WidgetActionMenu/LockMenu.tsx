import { IconButton, Tooltip } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';

import { LOCK_COLOR } from 'constants/bounding-box';
import { ReactComponent as LockIcon } from 'assets/icons/lock.svg';

import { updateWidget } from 'modules/Editor/store/infographSlice';
import { selectIsWidgetLocked } from 'modules/Editor/store/infographSelector';
import { selectActiveWidgetIds } from 'modules/Editor/store/widgetSelector';
import { useAppDispatch, useAppSelector } from 'modules/Editor/store/hooks';
import { Keyboard } from 'modules/common/components/Keyboard';

export const LockMenu = () => {
  const dispatch = useAppDispatch();
  const { t } = useTranslation('editor_toolbar');

  const activeWidgetIds = useAppSelector(selectActiveWidgetIds);
  const isLocked = useAppSelector(selectIsWidgetLocked(activeWidgetIds[0]));

  const lockText = t('lockBtn.label');
  const unlockText = t('unlockBtn.label');

  const label = isLocked ? unlockText : lockText;
  const isDisabled = activeWidgetIds.length !== 1;

  const lockWidget = () => {
    const updatedWidget = { widgetId: activeWidgetIds[0], widgetData: { isLocked: !isLocked } };
    dispatch(updateWidget(updatedWidget));
  };

  return (
    <Tooltip
      hasArrow
      placement='bottom'
      label={
        <>
          {label}
          <Keyboard shortcut={['ctrl', 'alt', 'L']} macOs={['âŒ˜', 'Opt', 'L']} />
        </>
      }
      bg='black'
    >
      <IconButton
        size='sm'
        aria-label={label}
        icon={<LockIcon style={{ stroke: isLocked ? LOCK_COLOR : '' }} />}
        onClick={lockWidget}
        isDisabled={isDisabled}
      />
    </Tooltip>
  );
};
