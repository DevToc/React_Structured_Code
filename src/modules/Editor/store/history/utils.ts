export const isEmpty = (value: any): boolean => {
  return (
    value === null ||
    value === undefined ||
    (typeof value === 'object' && !Object.values(value).some((v) => v !== null && typeof v !== 'undefined')) ||
    (typeof value === 'string' && value.length === 0)
  );
};

export const isNotEmpty = (value: any) => !isEmpty(value);
