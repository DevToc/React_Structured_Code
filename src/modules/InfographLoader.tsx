import { ReactNode, ReactElement, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useAppDispatch, useAppSelector, setActivePage } from './Editor/store';
import { setFontsLoaded } from './Editor/store/pageControlSlice';
import { filterUnloadedFamilies, loadFonts } from '../hooks/useFont';
import { findFontsInPage } from './Editor/store/infographSlice.helpers';
import { loadInfograph } from './Editor/store/infographSlice';
import widgetVersionManager from '../widgets/WidgetVersionManager';
import { getFullInfograph } from '../libs/e2api/get';
import { InfographState } from '../types/infographTypes';
import { setOriginalEditorId } from './Editor/store/editorSettingsSlice';
import { removeDeletedWidgets } from '../widgets/Widget.helpers';
import { selectActivePage } from './Editor/store/pageSelector';
import { selectInfograph } from './Editor/store/infographSelector';

interface InfographLoaderProps {
  children: ReactNode;
  infographState?: InfographState;
}

/**
 * Loads infograph to the store.
 *
 * If infographId param is available in the route, loads it from the backend api.
 *
 * If static local infograph is provided in prop, it'll load it.
 *
 * @param children
 * @param infographState
 * @constructor
 */
const InfographLoader = ({ children, infographState }: InfographLoaderProps): ReactElement => {
  const params = useParams();

  const { id: infographId, pageId, privateLinkId } = params || {};

  const dispatch = useAppDispatch();
  const infograph = useAppSelector(selectInfograph);
  const activePageId = useAppSelector(selectActivePage);

  useEffect(() => {
    const fetchId = infographId ?? privateLinkId;
    if (fetchId) {
      // Fetch infograph
      getFullInfograph(fetchId, !!privateLinkId).then((infograph) => {
        // console.debug('fetched infograph', infograph)
        const openPageId = pageId || infograph?.pageOrder[0] || '';

        // Load fonts asynchronously, to block page render before font fetching, use await loadFonts instead
        // TODO: hmm clean this up into single dispatch?
        // Set e1 uuid as a reference id
        !!infographId && dispatch(setOriginalEditorId(infographId));

        // Temporary: filter out deleted widgets from the Infograph
        // until they are removed from the infograph object returned from the BE (as they are not updated)
        const filteredInfograph = removeDeletedWidgets(infograph);

        // update widgets to latest version
        const updatedWidgets = widgetVersionManager.update(filteredInfograph.widgets);
        filteredInfograph.widgets = updatedWidgets;

        dispatch(loadInfograph(filteredInfograph));
        dispatch(setActivePage({ pageId: openPageId }));
        loadFonts(findFontsInPage(filteredInfograph, openPageId)).then(() => {
          dispatch(setFontsLoaded(true));
        });
      });
    } else if (infographState) {
      const pageId = infographState.pageOrder[0] || '';
      dispatch(loadInfograph(infographState));
      dispatch(setActivePage({ pageId }));
      //   TODO: improve this, not loading fonts here as this is used for integration tests
      //   loadFonts(findFontsInPage(infograph, openPageId));
    }
  }, [dispatch, infographId, infographState, pageId, privateLinkId]);

  useEffect(() => {
    // Handle multipage font loads
    // when the current active page changes
    // Skip the first page load
    if (!!activePageId && infograph?.pageOrder.length > 1 && infograph?.pageOrder[0] !== activePageId) {
      const fonts = findFontsInPage(infograph, activePageId);
      const unloadedFamilies = filterUnloadedFamilies(fonts);

      if (unloadedFamilies.length > 0) {
        dispatch(setFontsLoaded(false));
        loadFonts(unloadedFamilies).then(() => {
          dispatch(setFontsLoaded(true));
        });
      }
    }
  }, [dispatch, infograph, activePageId]);

  return <>{children}</>;
};

export { InfographLoader };
