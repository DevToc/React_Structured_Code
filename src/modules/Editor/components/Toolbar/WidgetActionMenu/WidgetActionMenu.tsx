import { Box, Flex, IconButton, Popover, PopoverContent, PopoverTrigger } from '@chakra-ui/react';

import { ToolbarDivider } from 'modules/common/components/Toolbar/ToolbarDivider';
import { ReactComponent as MoreIcon } from 'assets/icons/more.svg';
import { GroupWidgetMenu } from './GroupWidgetMenu';
import { FlowModeToggleButton } from './FlowModeToggleButton';
import { LayerMenu } from './LayerMenu/LayerMenu';
import { LockMenu } from './LockMenu';
import { DeleteMenu } from './DeleteMenu';
import { DuplicateMenu } from './DuplicateMenu';
import { ForwardedRef, forwardRef } from 'react';

const moreOptionsLabel = 'More Options';

interface WidgetActionMenuProps {
  // Set to TRUE to render the options in a dropdown
  useDropdown: boolean;
}

export const WidgetActionMenu = forwardRef(
  ({ useDropdown }: WidgetActionMenuProps, ref: ForwardedRef<HTMLDivElement>) => {
    return (
      <Box ref={ref}>
        {useDropdown ? (
          <Flex gap={2} alignItems='center'>
            <ToolbarDivider />
            <Popover placement='bottom-end' closeOnBlur={true}>
              {({ isOpen }) => (
                <>
                  <PopoverTrigger>
                    <IconButton
                      aria-label={moreOptionsLabel}
                      isActive={isOpen}
                      icon={<MoreIcon />}
                      size='sm'
                      variant='icon-btn-toolbar-option'
                      outline='none'
                    />
                  </PopoverTrigger>
                  <PopoverContent w='fit-content' boxShadow='md' padding={2}>
                    <ActionMenuContent />
                  </PopoverContent>
                </>
              )}
            </Popover>
          </Flex>
        ) : (
          <ActionMenuContent />
        )}
      </Box>
    );
  },
);

WidgetActionMenu.displayName = 'WidgetActionMenu';

export const ActionMenuContent = () => (
  <Flex gap={2} alignItems='center' overflowX='auto'>
    <FlowModeToggleButton />
    <LayerMenu />
    <ToolbarDivider />
    <GroupWidgetMenu />
    <DuplicateMenu />
    <LockMenu />
    <DeleteMenu />
  </Flex>
);
