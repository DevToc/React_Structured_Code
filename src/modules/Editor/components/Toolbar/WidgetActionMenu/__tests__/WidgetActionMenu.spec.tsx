import { renderWithRedux } from 'utils/test-utils.test';
import { InfographLoader } from 'modules/InfographLoader';
import { EMPTY_INFOGRAPH } from 'utils/loadSampleInfograph';
import { WidgetActionMenu } from '../WidgetActionMenu';

describe('components/Toolbar/WidgetActionMenu', () => {
  it('should render dropdown', () => {
    const { container } = renderWithRedux(
      <InfographLoader infographState={EMPTY_INFOGRAPH}>
        <WidgetActionMenu useDropdown={true} />
      </InfographLoader>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render all action buttons', () => {
    const { container } = renderWithRedux(
      <InfographLoader infographState={EMPTY_INFOGRAPH}>
        <WidgetActionMenu useDropdown={false} />
      </InfographLoader>,
    );

    expect(container).toMatchSnapshot();
  });
});
