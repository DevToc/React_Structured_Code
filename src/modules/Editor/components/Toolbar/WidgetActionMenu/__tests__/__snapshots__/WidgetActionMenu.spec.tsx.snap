// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Toolbar/WidgetActionMenu should render all action buttons 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1k2orlt"
    >
      <button
        aria-label="[BETA] Toggle Flow Mode - OFF"
        class="chakra-button css-9os3ml"
        data-testid="flowmode-toggle-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
        >
          cube_transparent.svg
        </svg>
      </button>
      <button
        aria-controls="popover-content-:rh:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="chakra-button css-1ofqz4h"
        data-testid="widget-layer-menu-trigger-button"
        disabled=""
        id="popover-trigger-:rh:"
        type="button"
      >
        Layer
      </button>
      <div
        class="chakra-popover__popper css-1qq679y"
        style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
      >
        <section
          class="chakra-popover__content css-1go3j4f"
          data-testid="widget-layer-menu-content"
          id="popover-content-:rh:"
          role="dialog"
          style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
          tabindex="-1"
        >
          <div
            class="chakra-button__group css-1bqlam1"
            role="group"
          >
            <button
              class="chakra-button css-7mlsyj"
              type="button"
            >
              <div
                class="css-tdm6l"
              >
                <svg>
                  moveFront.svg
                </svg>
                <p
                  class="chakra-text css-yt20lj"
                >
                  Move to Front
                </p>
              </div>
              <div
                class="css-k008qs"
              >
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  ctrl
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  shift
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  &gt;
                </kbd>
              </div>
            </button>
            <button
              class="chakra-button css-7mlsyj"
              type="button"
            >
              <div
                class="css-tdm6l"
              >
                <svg>
                  moveForwards.svg
                </svg>
                <p
                  class="chakra-text css-yt20lj"
                >
                  Move Forward
                </p>
              </div>
              <div
                class="css-k008qs"
              >
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  ctrl
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  &gt;
                </kbd>
              </div>
            </button>
            <button
              class="chakra-button css-7mlsyj"
              type="button"
            >
              <div
                class="css-tdm6l"
              >
                <svg>
                  moveBackwards.svg
                </svg>
                <p
                  class="chakra-text css-yt20lj"
                >
                  Move Backwards
                </p>
              </div>
              <div
                class="css-k008qs"
              >
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  ctrl
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  &lt;
                </kbd>
              </div>
            </button>
            <button
              class="chakra-button css-7mlsyj"
              type="button"
            >
              <div
                class="css-tdm6l"
              >
                <svg>
                  moveBack.svg
                </svg>
                <p
                  class="chakra-text css-yt20lj"
                >
                  Move to Back
                </p>
              </div>
              <div
                class="css-k008qs"
              >
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  ctrl
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  shift
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  +
                </kbd>
                <kbd
                  class="chakra-kbd css-2fship"
                >
                  &lt;
                </kbd>
              </div>
            </button>
          </div>
        </section>
      </div>
      <hr
        aria-orientation="vertical"
        class="chakra-divider css-16aks9g"
      />
      <button
        class="chakra-button css-1rt59zj"
        data-testid="group-widget-menu"
        disabled=""
        type="button"
      >
        Group
      </button>
      <button
        aria-label="Duplicate"
        class="chakra-button css-108yenj"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
        >
          copy.svg
        </svg>
      </button>
      <button
        aria-label="Lock"
        class="chakra-button css-108yenj"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
        >
          lock.svg
        </svg>
      </button>
      <button
        aria-label="Delete widget"
        class="chakra-button css-108yenj"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
        >
          trash.svg
        </svg>
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`components/Toolbar/WidgetActionMenu should render dropdown 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-1p3wbw4"
    >
      <hr
        aria-orientation="vertical"
        class="chakra-divider css-16aks9g"
      />
      <button
        aria-controls="popover-content-:r1:"
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-label="More Options"
        class="chakra-button css-1dzjj48"
        id="popover-trigger-:r1:"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
        >
          more.svg
        </svg>
      </button>
      <div
        class="chakra-popover__popper css-1qq679y"
        style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
      >
        <section
          class="chakra-popover__content css-5sqc17"
          id="popover-content-:r1:"
          role="dialog"
          style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
          tabindex="-1"
        >
          <div
            class="css-1k2orlt"
          >
            <button
              aria-label="[BETA] Toggle Flow Mode - OFF"
              class="chakra-button css-9os3ml"
              data-testid="flowmode-toggle-button"
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
              >
                cube_transparent.svg
              </svg>
            </button>
            <button
              aria-controls="popover-content-:r5:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="chakra-button css-1ofqz4h"
              data-testid="widget-layer-menu-trigger-button"
              disabled=""
              id="popover-trigger-:r5:"
              type="button"
            >
              Layer
            </button>
            <div
              class="chakra-popover__popper css-1qq679y"
              style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
            >
              <section
                class="chakra-popover__content css-1go3j4f"
                data-testid="widget-layer-menu-content"
                id="popover-content-:r5:"
                role="dialog"
                style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
                tabindex="-1"
              >
                <div
                  class="chakra-button__group css-1bqlam1"
                  role="group"
                >
                  <button
                    class="chakra-button css-7mlsyj"
                    type="button"
                  >
                    <div
                      class="css-tdm6l"
                    >
                      <svg>
                        moveFront.svg
                      </svg>
                      <p
                        class="chakra-text css-yt20lj"
                      >
                        Move to Front
                      </p>
                    </div>
                    <div
                      class="css-k008qs"
                    >
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        ctrl
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        shift
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        &gt;
                      </kbd>
                    </div>
                  </button>
                  <button
                    class="chakra-button css-7mlsyj"
                    type="button"
                  >
                    <div
                      class="css-tdm6l"
                    >
                      <svg>
                        moveForwards.svg
                      </svg>
                      <p
                        class="chakra-text css-yt20lj"
                      >
                        Move Forward
                      </p>
                    </div>
                    <div
                      class="css-k008qs"
                    >
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        ctrl
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        &gt;
                      </kbd>
                    </div>
                  </button>
                  <button
                    class="chakra-button css-7mlsyj"
                    type="button"
                  >
                    <div
                      class="css-tdm6l"
                    >
                      <svg>
                        moveBackwards.svg
                      </svg>
                      <p
                        class="chakra-text css-yt20lj"
                      >
                        Move Backwards
                      </p>
                    </div>
                    <div
                      class="css-k008qs"
                    >
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        ctrl
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        &lt;
                      </kbd>
                    </div>
                  </button>
                  <button
                    class="chakra-button css-7mlsyj"
                    type="button"
                  >
                    <div
                      class="css-tdm6l"
                    >
                      <svg>
                        moveBack.svg
                      </svg>
                      <p
                        class="chakra-text css-yt20lj"
                      >
                        Move to Back
                      </p>
                    </div>
                    <div
                      class="css-k008qs"
                    >
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        ctrl
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        shift
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        +
                      </kbd>
                      <kbd
                        class="chakra-kbd css-2fship"
                      >
                        &lt;
                      </kbd>
                    </div>
                  </button>
                </div>
              </section>
            </div>
            <hr
              aria-orientation="vertical"
              class="chakra-divider css-16aks9g"
            />
            <button
              class="chakra-button css-1rt59zj"
              data-testid="group-widget-menu"
              disabled=""
              type="button"
            >
              Group
            </button>
            <button
              aria-label="Duplicate"
              class="chakra-button css-108yenj"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
              >
                copy.svg
              </svg>
            </button>
            <button
              aria-label="Lock"
              class="chakra-button css-108yenj"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
              >
                lock.svg
              </svg>
            </button>
            <button
              aria-label="Delete widget"
              class="chakra-button css-108yenj"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
              >
                trash.svg
              </svg>
            </button>
          </div>
        </section>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
